### PR TITLE
Create SubscriptionRequest

### PR DESCRIPTION
**Pull request by gbirke**
Create SubscriptionRequest from Silex HTTP request data (GET, POST,
Cookies).

### CODE REVIEW

File: src/UseCases/AddSubscription/SubscriptionRequest.php [hunk 1]
```diff
@@ -91,4 +91,27 @@ public function setWikilogin( bool $wikilogin ) {
...
+	 */
+	public function setWikiloginFromValues( array $values ) {
+		$trueValues = ['yes', '1'];
+		$falseValues = ['no', '0'];
+		$matchingValues = array_intersect( $values, array_merge( $trueValues, $falseValues ) );
+		$wikilogin = in_array( array_shift( $matchingValues ), $trueValues );
+		$this->setWikilogin( $wikilogin );
+	}
+
+	public static function createFromArray( array $values ): SubscriptionRequest {
```
Comment (JeroenDeDauw) [inline src/UseCases/AddSubscription/SubscriptionRequest.php]: This is presentation related code on the wrong side of the boundary.

File: src/UseCases/AddSubscription/SubscriptionRequest.php [hunk 2]
```diff
@@ -91,4 +91,27 @@ public function setWikilogin( bool $wikilogin ) {
...
+	 * @param array $values
+	 */
+	public function setWikiloginFromValues( array $values ) {
+		$trueValues = ['yes', '1'];
+		$falseValues = ['no', '0'];
+		$matchingValues = array_intersect( $values, array_merge( $trueValues, $falseValues ) );
+		$wikilogin = in_array( array_shift( $matchingValues ), $trueValues );
+		$this->setWikilogin( $wikilogin );
+	}
+
```
Comment (JeroenDeDauw) [inline src/UseCases/AddSubscription/SubscriptionRequest.php]: In such a case you can use `self` as type hint

File: src/UseCases/AddSubscription/SubscriptionRequest.php [hunk 3]
```diff
@@ -91,4 +91,27 @@ public function setWikilogin( bool $wikilogin ) {
...
+	public static function createFromArray( array $values ): SubscriptionRequest {
+		$instance = new self();
+		foreach ( $values as $key => $value ) {
+			$accessor = 'set' . ucfirst( $key );
+			if ( is_callable( [$instance, $accessor ] ) ) {
+				call_user_func( [$instance, $accessor ], $value );
```
Comment (JeroenDeDauw) [inline src/UseCases/AddSubscription/SubscriptionRequest.php]: I really do not like this kind of magic. I understand that doing

```php
$object->setFoo( $params['foo'] );
$object->setBar( $params['bar'] );
// ...
```

is boring. However it's very easy to understand and to maintain. It does not mess up static code analysis, and it does not make me wonder if there is some way this code could be exploited by a malicious agent.

Comment (JeroenDeDauw) [inline src/UseCases/AddSubscription/SubscriptionRequest.php]: This pattern also prevents using composition without adding shortcut methods. Looks like `SubscriptionRequest` could contain a `PesonName` value object with first and last name and an `Address` value object with street address, postcode and city.

File: app/routes.php [hunk 4]
```diff
@@ -48,8 +48,11 @@ function( Application $app, $pageName ) use ( $ffFactory ) {
 	'contact/subscribe',
 	function( Request $request ) use ( $ffFactory ) {
 		$useCase = $ffFactory->newAddSubscriptionUseCase();
-		// TODO: Create SubscriptionRequest from $request params
-		$subscriptionRequest = new SubscriptionRequest();
+		$subscriptionRequest = SubscriptionRequest::createFromArray( $request->request->all() );
```
Comment (JeroenDeDauw) [inline app/routes.php]: The new code here appears to be untested

File: tests/Unit/UseCases/AddSubscription/SubscriptionRequestTest.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+		] );
+		$this->assertEquals( $subscriptionRequest->getFirstName(), 'Nyan' );
+		$this->assertEquals( $subscriptionRequest->getLastName(), 'Cat' );
+		$this->assertEquals( $subscriptionRequest->getSalutation(), 'Herr' );
+		$this->assertEquals( $subscriptionRequest->getTitle(), 'Prof. Dr. Dr.' );
+		$this->assertEquals( $subscriptionRequest->getAddress(), 'Awesome Way 1' );
+		$this->assertEquals( $subscriptionRequest->getCity(), 'Berlin' );
+		$this->assertEquals( $subscriptionRequest->getPostcode(), '12345' );
+		$this->assertEquals( $subscriptionRequest->getEmail(), 'me@nyancat.wtf' );
+		$this->assertEquals( $subscriptionRequest->getWikilogin(), true );
```
Comment (JeroenDeDauw) [inline tests/Unit/UseCases/AddSubscription/SubscriptionRequestTest.php]: assertSame would be more appropriate for all of these

File: tests/Unit/UseCases/AddSubscription/SubscriptionRequestTest.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ * @licence GNU GPL v2+
+ * @author Gabriel Birke < gabriel.birke@wikimedia.de >
+ */
+class SubscriptionRequestTest extends \PHPUnit_Framework_TestCase {
+	public function testSubscriptionRequestCanBeInitializedFromArray() {
+		$subscriptionRequest = SubscriptionRequest::createFromArray( [
+			'firstName' => 'Nyan',
+			'lastName' => 'Cat',
+			'salutation' => 'Herr',
+			'title' => 'Prof. Dr. Dr.',
```
Comment (JeroenDeDauw) [inline tests/Unit/UseCases/AddSubscription/SubscriptionRequestTest.php]: Took me till seeing this to understand what title meant. In the request TDO this was certainly not clear to me. This is another reason to not use the dynamic mapping magic: it forces the setter names to match those of the presentation interface.

File: src/UseCases/AddSubscription/SubscriptionRequest.php (hunk 1 above)
Comment (JeroenDeDauw) [inline src/UseCases/AddSubscription/SubscriptionRequest.php]: Fixed by #46

File: src/UseCases/AddSubscription/SubscriptionRequest.php (hunk 3 above)
Comment (JeroenDeDauw) [inline src/UseCases/AddSubscription/SubscriptionRequest.php]: Replaced in #46

File: app/routes.php (hunk 4 above)
Comment (JeroenDeDauw) [inline app/routes.php]: Partially addressed by #46. Even found a bug when writing the test :)

File: tests/Unit/UseCases/AddSubscription/SubscriptionRequestTest.php (hunk 5 above)
Comment (JeroenDeDauw) [inline tests/Unit/UseCases/AddSubscription/SubscriptionRequestTest.php]: Made obsolete by #46

**Comment by gbirke** [comment]
Since I've accepted the edits by Jeroen, this can now be merged.

**Pull request merged**